plugins/auth-backend-module-pinniped-provider/src/authenticator.ts

```diff
@@ -54,21 +57,75 @@ const rfc8693TokenExchange = async ({
 };
 
 /** @public */
-export const pinnipedAuthenticator = createOAuthAuthenticator({
-  defaultProfileTransform: async (_r, _c) => ({ profile: {} }),
-  async initialize({ callbackUrl, config }) {
+export class PinnipedStrategyFactory {
+  private readonly callbackUrl: string;
+  private readonly config: Config;
+  private strategyPromise: Promise<{
+    providerStrategy: OidcStrategy<{ tokenset: TokenSet }, BaseClient>;
+    client: BaseClient;
+  }>;
+
+  private cachedPromise?: Promise<{
+    providerStrategy: OidcStrategy<{ tokenset: TokenSet }, BaseClient>;
+    client: BaseClient;
+  }>;
+  private cachedPromiseExpiry?: Date;
+
+  constructor(callbackUrl: string, config: Config) {
+    this.callbackUrl = callbackUrl;
+    this.config = config;
+    this.strategyPromise = this.buildStrategy();
+  }
+
+  public async getStrategy(): Promise<{
+    providerStrategy: OidcStrategy<{ tokenset: TokenSet }, BaseClient>;
+    client: BaseClient;
+  }> {
+    if (this.cachedPromise) {
+      if (
+        this.cachedPromiseExpiry &&
+        DateTime.fromJSDate(this.cachedPromiseExpiry) > DateTime.local()
+      ) {
+        return this.cachedPromise;
+      }
+      // cachedPromise has expired, remove promise from cache and regenerate strategy
+      this.strategyPromise = this.buildStrategy();
+      delete this.cachedPromise;
+    }
+
+    try {
+      // if strategy is generated successfully, save it to cache
+      await this.strategyPromise;
+      this.cachedPromise = this.strategyPromise;
+      this.cachedPromiseExpiry = DateTime.utc()
+        .plus({ seconds: 3600 })
```

Comment:
I think this 1 hour TTL for metadata should at least be extracted to a module-level constant. Might even make sense to make it configurable, but I would wait for a later PR to see if people genuinely want that

Reply:
Ive extracted and defined a constant in the authenticator file and added the export to the modules index.ts file. I hope thats what you meant by module-level constant? Ok we can definitely make it configurable if anyone prefers but ill wait. I took the value from the backend-session-expiration constant that was once used in the auth-backend, and also from the TokenFactory that i saw uses a similar mechanism to cache. I have it in as a placeholder value for now but definitely open to suggestions if someone wanted some other value for cache expiration.